File: xllm/core/kernels/mlu/tmo_kernel/triton_kernel/causal_conv1d_update_decode.py
# Copyright 2026 The xLLM Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://github.com/jd-opensource/xllm/blob/main/LICENSE
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# This kernel is adapted from vLLM's FLA Triton ops:
# https://github.com/vllm-project/vllm/tree/v0.18.0/vllm/model_executor/layers/fla/ops
# Upstream license: Apache License, Version 2.0.
# Modified for xLLM MLU TMO integration.

import triton
import triton.language as tl


@triton.jit
def tmo_causal_conv1d_update_decode_kernel(
    # Pointers to tensors
    x_ptr,  # (batch, dim, seqlen) or (dim, cu_seqlen)
    weight_ptr,  # (dim, width)
    bias_ptr,  # (dim,) or nullptr
    conv_state_ptr,  # (num_cache_lines, dim, state_len)
    conv_state_indices_ptr,  # (batch,) int32
    num_accepted_tokens_ptr,  # (batch,) int32 or nullptr
    query_start_loc_ptr,  # (batch+1,) int32 or nullptr
    block_idx_last_scheduled_token_ptr,  # (batch,) int32 or nullptr
    initial_state_idx_ptr,  # (batch,) int32 or nullptr
    out_ptr,  # (batch, dim, seqlen) or (dim, cu_seqlen)
    # Dimensions (runtime)
    batch,
    num_cache_lines,
    # Dimensions (constexpr for AOT specialization)
    dim: tl.constexpr,
    seqlen: tl.constexpr,
    state_len: tl.constexpr,
    # Strides (runtime)
    stride_x_seq,
    stride_x_dim,
    stride_x_token,
    stride_w_dim,
    stride_w_width,
    stride_istate_seq,
    stride_istate_dim,
    stride_istate_tok,
    stride_state_indices,
    stride_o_seq,
    stride_o_dim,
    stride_o_token,
    # Others (runtime)
    pad_slot_id,
    # Meta-parameters (constexpr for AOT specialization)
    HAS_BIAS: tl.constexpr,
    KERNEL_WIDTH: tl.constexpr,
    SILU_ACTIVATION: tl.constexpr,
    IS_VARLEN: tl.constexpr,
    IS_APC_ENABLED: tl.constexpr,
    IS_SPEC_DECODING: tl.constexpr,
    NP2_STATELEN: tl.constexpr,
    USE_PAD_SLOT: tl.constexpr,
    BD: tl.constexpr,
    BW: tl.constexpr,
):
    # ruff: noqa: E501
    # Grid layout: program_id(0) = feature block, program_id(1) = sequence
    # This follows the torch_mlu_ops convention for better MLU parallelism.
    i_d = tl.program_id(0) * BD + tl.arange(0, BD)  # [BD] feature indices
    i_n = tl.program_id(1)  # sequence index

    if i_n >= batch:
        return

    # APC: load initial_state_idx and block_idx_last_scheduled_token
    if IS_APC_ENABLED:
        conv_state_init = tl.load(initial_state_idx_ptr + i_n)
        current_last_index = tl.load(block_idx_last_scheduled_token_ptr + i_n)
    else:
        conv_state_init = 0
        current_last_index = 0

    # Load conv_state_index for this sequence
    conv_states_input_coord = tl.load(
        conv_state_indices_ptr + i_n * stride_state_indices + conv_state_init
    ).to(tl.int64)

    # Skip padded entries
    if USE_PAD_SLOT:
        if conv_states_input_coord == pad_slot_id:
            return

    # Varlen: compute sequence start/end from query_start_loc
    if IS_VARLEN:
        query_start_index = tl.load(query_start_loc_ptr + i_n).to(tl.int64)
        query_end_index = tl.load(query_start_loc_ptr + (i_n + 1)).to(tl.int64)
        # revise state_len and seqlen for this specific sequence
        cur_state_len = state_len - (seqlen - (query_end_index - query_start_index))
        cur_seqlen = query_end_index - query_start_index
        x_offset = query_start_index * stride_x_token
        o_offset = query_start_index * stride_o_token
    else:
        query_start_index = i_n * seqlen
        query_end_index = query_start_index + seqlen
        cur_state_len = state_len
        cur_seqlen = seqlen
        x_offset = i_n * stride_x_seq
        o_offset = i_n * stride_o_seq

    if query_start_index == query_end_index:
        return

    # Speculative decoding: compute conv_state token offset
    if IS_SPEC_DECODING:
        conv_state_token_offset = (
            tl.load(num_accepted_tokens_ptr + i_n).to(tl.int64) - 1
        )
    else:
        conv_state_token_offset = 0

    # STEP 1: READ init_state prior tokens from conv_state
    # conv_states_base points to the feature slice for this sequence
    conv_states_base = (
        conv_state_ptr
        + (conv_states_input_coord * stride_istate_seq)
        + (i_d * stride_istate_dim)
    )
    mask_feat = i_d < dim

    prior_tokens = conv_states_base + conv_state_token_offset * stride_istate_tok
    if KERNEL_WIDTH >= 2:
        col0 = tl.load(prior_tokens + 0 * stride_istate_tok, mask_feat, 0.0).to(tl.float32)
    if KERNEL_WIDTH >= 3:
        col1 = tl.load(prior_tokens + 1 * stride_istate_tok, mask_feat, 0.0).to(tl.float32)
    if KERNEL_WIDTH >= 4:
        col2 = tl.load(prior_tokens + 2 * stride_istate_tok, mask_feat, 0.0).to(tl.float32)
    if KERNEL_WIDTH >= 5:
        col3 = tl.load(prior_tokens + 3 * stride_istate_tok, mask_feat, 0.0).to(tl.float32)
    if KERNEL_WIDTH >= 6:
        col4 = tl.load(prior_tokens + 4 * stride_istate_tok, mask_feat, 0.0).to(tl.float32)

    # STEP 2: Build new_conv_state (shift + append) and write back
    idx_tokens = tl.arange(0, NP2_STATELEN)  # [NP2_STATELEN]

    # Load existing conv_state for shift
    conv_state_ptrs_source = (
        conv_state_ptr
        + (conv_states_input_coord * stride_istate_seq)
        + conv_state_token_offset * stride_istate_tok
        + (i_d * stride_istate_dim)[None, :]
        + ((idx_tokens + (1 if IS_SPEC_DECODING else cur_seqlen)) * stride_istate_tok)[
            :, None
        ]
    )  # [NP2_STATELEN, BD]
    mask_shift = (
        (conv_states_input_coord < num_cache_lines)
        & ((idx_tokens + cur_seqlen) < cur_state_len)[:, None]
        & (i_d < dim)[None, :]
    )
    conv_state_shifted = tl.load(conv_state_ptrs_source, mask_shift, other=0.0)

    # Load x values to fill the new positions
    VAL = cur_state_len - cur_seqlen
    x_base = x_ptr + x_offset + (i_d * stride_x_dim)  # [BD]

    x_ptrs = (
        x_base[None, :] + ((idx_tokens - VAL) * stride_x_token)[:, None]
    )  # [NP2_STATELEN, BD]
    mask_x = (
        (idx_tokens - VAL >= 0)[:, None]
        & (idx_tokens - VAL < cur_seqlen)[:, None]
        & (i_d < dim)[None, :]
    )
    loaded_x = tl.load(x_ptrs, mask_x, 0.0)
    tl.debug_barrier()

    # Merge: keep shifted state where mask_shift is True, use x otherwise
    new_conv_state = tl.where(mask_shift, conv_state_shifted, loaded_x)

    # Determine target cache slot (for APC, may be different from source)
    conv_states_offset = tl.load(
        conv_state_indices_ptr + i_n * stride_state_indices + current_last_index
    ).to(tl.int64)
    conv_state_ptrs_target = (
        conv_state_ptr
        + (conv_states_offset * stride_istate_seq)
        + (i_d * stride_istate_dim)
    )[None, :] + (
        idx_tokens * stride_istate_tok
    )[:, None]
    mask_target = (idx_tokens < cur_state_len)[:, None] & (i_d < dim)[None, :]
    tl.store(conv_state_ptrs_target, new_conv_state, mask_target)

    # STEP 3: Initialize accumulator with bias (or zero)
    if HAS_BIAS:
        acc_preload = tl.load(bias_ptr + i_d, mask=mask_feat, other=0.0).to(tl.float32)
    else:
        acc_preload = tl.zeros((BD,), dtype=tl.float32)

    # STEP 4: Pre-load weight columns
    w_base = weight_ptr + (i_d * stride_w_dim)  # [BD]
    if KERNEL_WIDTH >= 2:
        w_col0 = tl.load(w_base + 0 * stride_w_width, mask_feat, 0.0).to(tl.float32)
        w_col1 = tl.load(w_base + 1 * stride_w_width, mask_feat, 0.0).to(tl.float32)
    if KERNEL_WIDTH >= 3:
        w_col2 = tl.load(w_base + 2 * stride_w_width, mask_feat, 0.0).to(tl.float32)
    if KERNEL_WIDTH >= 4:
        w_col3 = tl.load(w_base + 3 * stride_w_width, mask_feat, 0.0).to(tl.float32)
    if KERNEL_WIDTH >= 5:
        w_col4 = tl.load(w_base + 4 * stride_w_width, mask_feat, 0.0).to(tl.float32)
    if KERNEL_WIDTH >= 6:
        w_col5 = tl.load(w_base + 5 * stride_w_width, mask_feat, 0.0).to(tl.float32)

    # STEP 5: Compute convolution output for each token in the sequence
    for idx_token in tl.range(seqlen):
        acc = acc_preload

        # Compute depthwise conv1d: sum of (prior_state[i] * weight[i]) for i in KERNEL_WIDTH
        # The last column always uses the current x token
        for j in tl.static_range(KERNEL_WIDTH):
            if KERNEL_WIDTH == 2:
                if j == 0:
                    matrix_w = w_col0
                    matrix_x = col0
                elif j == 1:
                    matrix_w = w_col1
                    matrix_x = tl.load(
                        x_base + idx_token * stride_x_token, mask_feat, 0.0
                    ).to(tl.float32)
            elif KERNEL_WIDTH == 3:
                if j == 0:
                    matrix_w = w_col0
                    matrix_x = col0
                elif j == 1:
                    matrix_w = w_col1
                    matrix_x = col1
                elif j == 2:
                    matrix_w = w_col2
                    matrix_x = tl.load(
                        x_base + idx_token * stride_x_token, mask_feat, 0.0
                    ).to(tl.float32)
            elif KERNEL_WIDTH == 4:
                if j == 0:
                    matrix_w = w_col0
                    matrix_x = col0
                elif j == 1:
                    matrix_w = w_col1
                    matrix_x = col1
                elif j == 2:
                    matrix_w = w_col2
                    matrix_x = col2
                elif j == 3:
                    matrix_w = w_col3
                    matrix_x = tl.load(
                        x_base + idx_token * stride_x_token, mask_feat, 0.0
                    ).to(tl.float32)
            elif KERNEL_WIDTH == 5:
                if j == 0:
                    matrix_w = w_col0
                    matrix_x = col0
                elif j == 1:
                    matrix_w = w_col1
                    matrix_x = col1
                elif j == 2:
                    matrix_w = w_col2
                    matrix_x = col2
                elif j == 3:
                    matrix_w = w_col3
                    matrix_x = col3
                elif j == 4:
                    matrix_w = w_col4
                    matrix_x = tl.load(
                        x_base + idx_token * stride_x_token, mask_feat, 0.0
                    ).to(tl.float32)
            elif KERNEL_WIDTH == 6:
                if j == 0:
                    matrix_w = w_col0
                    matrix_x = col0
                elif j == 1:
                    matrix_w = w_col1
                    matrix_x = col1
                elif j == 2:
                    matrix_w = w_col2
                    matrix_x = col2
                elif j == 3:
                    matrix_w = w_col3
                    matrix_x = col3
                elif j == 4:
                    matrix_w = w_col4
                    matrix_x = col4
                elif j == 5:
                    matrix_w = w_col5
                    matrix_x = tl.load(
                        x_base + idx_token * stride_x_token, mask_feat, 0.0
                    ).to(tl.float32)

            acc += matrix_x * matrix_w  # [BD] float32 accumulation

        # Update history columns (shift window)
        if KERNEL_WIDTH == 2:
            col0 = matrix_x
        elif KERNEL_WIDTH == 3:
            col0 = col1
            col1 = matrix_x
        elif KERNEL_WIDTH == 4:
            col0 = col1
            col1 = col2
            col2 = matrix_x
        elif KERNEL_WIDTH == 5:
            col0 = col1
            col1 = col2
            col2 = col3
            col3 = matrix_x
        elif KERNEL_WIDTH == 6:
            col0 = col1
            col1 = col2
            col2 = col3
            col3 = col4
            col4 = matrix_x

        # Apply SiLU activation: x * sigmoid(x) = x / (1 + exp(-x))
        if SILU_ACTIVATION:
            acc = acc / (1 + tl.exp(-acc))

        # Store output
        mask_1d = i_d < dim
        o_ptrs = (
            out_ptr + o_offset + idx_token * stride_o_token + (i_d * stride_o_dim)
        )
        tl.store(o_ptrs, acc, mask=mask_1d)

File: xllm/core/kernels/mlu/causal_conv1d_update_decode.cpp
/* Copyright 2026 The xLLM Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    https://github.com/jd-opensource/xllm/blob/main/LICENSE

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include "causal_conv1d_update_decode.h"

#include <cnrt.h>
#include <framework/core/MLUStream.h>
#include <glog/logging.h>

#include <algorithm>
#include <unordered_map>

#include "kernels/mlu/mlu_ops_api.h"
#include "kernels/mlu/utils.h"

namespace xllm {
namespace kernel {
namespace mlu {

torch::Tensor causal_conv1d_update_decode(
    const torch::Tensor& x,
    torch::Tensor& conv_state,
    const torch::Tensor& weight,
    const std::optional<torch::Tensor>& bias_opt,
    const torch::Tensor& conv_state_indices,
    int32_t pad_slot_id,
    const std::optional<torch::Tensor>& query_start_loc_opt,
    int32_t max_query_len,
    const std::optional<torch::Tensor>& num_accepted_tokens_opt,
    const std::optional<torch::Tensor>& block_idx_last_scheduled_token_opt,
    const std::optional<torch::Tensor>& initial_state_idx_opt) {
  bool unsqueeze = (x.dim() == 2);
  torch::Tensor x_input = x;
  if (unsqueeze) {
    x_input = x.unsqueeze(-1);
  }

  int32_t batch = static_cast<int32_t>(conv_state_indices.size(0));
  int32_t dim = static_cast<int32_t>(x_input.size(1));
  int32_t seqlen = static_cast<int32_t>(x_input.size(2));
  int32_t width = static_cast<int32_t>(weight.size(1));
  int32_t state_len = width - 1;
  int32_t num_cache_lines = static_cast<int32_t>(conv_state.size(0));

  // Create output tensor with same shape and dtype as x
  torch::Tensor out = torch::empty_like(x_input);

  // Strides for x (batch, dim, seqlen)
  int32_t stride_x_seq = static_cast<int32_t>(x_input.stride(0));
  int32_t stride_x_dim = static_cast<int32_t>(x_input.stride(1));
  int32_t stride_x_token = static_cast<int32_t>(x_input.stride(2));
  // Strides for weight (dim, width)
  int32_t stride_w_dim = static_cast<int32_t>(weight.stride(0));
  int32_t stride_w_width = static_cast<int32_t>(weight.stride(1));
  // Strides for conv_state (num_cache_lines, dim, state_len)
  int32_t stride_istate_seq = static_cast<int32_t>(conv_state.stride(0));
  int32_t stride_istate_dim = static_cast<int32_t>(conv_state.stride(1));
  int32_t stride_istate_tok = static_cast<int32_t>(conv_state.stride(2));
  // Strides for conv_state_indices (batch)
  int32_t stride_state_indices =
      static_cast<int32_t>(conv_state_indices.stride(0));
  // Strides for out (batch, dim, seqlen)
  int32_t stride_o_seq = static_cast<int32_t>(out.stride(0));
  int32_t stride_o_dim = static_cast<int32_t>(out.stride(1));
  int32_t stride_o_token = static_cast<int32_t>(out.stride(2));

  // Data pointers
  void* x_ptr = x_input.data_ptr();
  void* weight_ptr = weight.data_ptr();
  void* bias_ptr = bias_opt.has_value() ? bias_opt->data_ptr() : nullptr;
  void* conv_state_ptr = conv_state.data_ptr();
  void* conv_state_indices_ptr = conv_state_indices.data_ptr();
  void* num_accepted_tokens_ptr = num_accepted_tokens_opt.has_value()
                                      ? num_accepted_tokens_opt->data_ptr()
                                      : nullptr;
  void* query_start_loc_ptr = query_start_loc_opt.has_value()
                                  ? query_start_loc_opt->data_ptr()
                                  : nullptr;
  void* block_idx_last_scheduled_token_ptr =
      block_idx_last_scheduled_token_opt.has_value()
          ? block_idx_last_scheduled_token_opt->data_ptr()
          : nullptr;
  void* initial_state_idx_ptr = initial_state_idx_opt.has_value()
                                    ? initial_state_idx_opt->data_ptr()
                                    : nullptr;
  void* out_ptr = out.data_ptr();

  bool has_bias = bias_opt.has_value();

  constexpr int32_t kBd = 8;
  int32_t num_feature_blocks = (dim + kBd - 1) / kBd;
  cnrtDim3_t dim_block = {static_cast<uint32_t>(num_feature_blocks),
                          static_cast<uint32_t>(batch),
                          1};

  auto queue = torch_mlu::getCurMLUStream();

  // algo_id: select pre-compiled kernel variant based on dim value
  static const std::unordered_map<int32_t, int32_t> kDimToAlgoId = {
      {384, 0},
      {512, 1},
      {640, 2},
      {768, 3},
      {1024, 4},
      {1280, 5},
      {1536, 6},
      {2048, 7},
      {2560, 8},
      {3072, 9},
      {4096, 10},
      {5120, 11},
      {6144, 12},
      {8192, 13},
      {10240, 14},
      {12288, 15},
  };
  int32_t algo_id = lookup_algo_id(kDimToAlgoId, dim, /*dim_name=*/"dim");

  tmo_causal_conv1d_update_decode_kernel(queue,
                                         &dim_block,
                                         x_ptr,
                                         weight_ptr,
                                         bias_ptr,
                                         conv_state_ptr,
                                         conv_state_indices_ptr,
                                         num_accepted_tokens_ptr,
                                         query_start_loc_ptr,
                                         block_idx_last_scheduled_token_ptr,
                                         initial_state_idx_ptr,
                                         out_ptr,
                                         batch,
                                         num_cache_lines,
                                         stride_x_seq,
                                         stride_x_dim,
                                         stride_x_token,
                                         stride_w_dim,
                                         stride_w_width,
                                         stride_istate_seq,
                                         stride_istate_dim,
                                         stride_istate_tok,
                                         stride_state_indices,
                                         stride_o_seq,
                                         stride_o_dim,
                                         stride_o_token,
                                         pad_slot_id,
                                         algo_id);

  if (unsqueeze) {
    out = out.squeeze(-1);
  }
  return out;
}

}  // namespace mlu
}  // namespace kernel
}  // namespace xllm
